src/main/combatLogParser.ts

```diff
@@ -0,0 +1,434 @@
+import { glob } from "glob";
+import path from "path";
+import fs from "fs";
+import { EventEmitter } from "stream";
+import { setInterval, clearInterval } from 'timers';
+
+const tail = require('tail').Tail;
+
+/**
+ * Type that defines the structure of a combat log file handler,
+ * which is used to monitor a specific log directory.
+ */
+type CombatLogMonitorHandlerType = {
+    path: string,
+    wowFlavour: string, // Read from the '.flavor.info' file from WoW
+    currentLogFile?: string,
+    watchInterval?: any,
+    tail?: any,
+};
+
+type CombatLogParserOptionsType = {
+    dataTimeout: number,
+};
+
+/**
+ * A just-in-time parsed line from the WoW combat log
+ *
+ * The object is constructed from the original combat log line and will
+ * parse log line arguments incrementally as they are requested, that is,
+ * a log line like CHALLENGE_MODE_START which has few arguments won't see
+ * much performance gain over parsing everything, but COMBATANT_INFO will
+ * due to its absurdly long list of arguments that we most often won't use.
+ */
+class LogLine {
+    // Current parsing position in the original line
+    private _linePosition = 0;
+    // Length of the original line to avoid reevaluating it
+    // many times.
+    private _lineLength = 0;
+
+    // Multi-dimensional array of arguments
+    // Example: 'ARENA_MATCH_START', '2547', '33', '2v2', '1'
+    private _args: any[] = [];
+
+    // Length of this.args to avoid evaluating this.args.length
+    // may times.
+    private _argsListLen = 0;
+
+    // Timestamp in string format, as-is, from the log
+    // Example: '8/3 22:09:58.548'
+    public timestamp: string = '';
+
+    constructor (
+        // Original line as it came from the log
+        public original: string
+    ) {
+        this._lineLength = this.original.length;
+
+        // Combat log line always has '<timestamp>  <line>' format,
+        // that is, two spaces between ts and line.
+        this._linePosition = this.original.indexOf('  ') + 2;
+        this.timestamp = this.original.substring(0, this._linePosition - 2);
+
+        // Parse the first argument, which is the event type and will always
+        // be needed.
+        this.parseLogArg(1);
+    }
+
+    arg(index: number): any {
+        if (!this._args || index >= this._argsListLen) {
+            const maxsplit = Math.max(index + 1, this._argsListLen);
+            this.parseLogArg(maxsplit);
+        }
+
+        return this._args[index];
+    }
+
+    /**
+     * Parse the timestamp from a log line and create a Date value from it
+     */
+    date(): Date {
+        // Split the line by any delimiter that isn't a number,
+        // convert them to actual numbers and reverse the array.
+        const timeParts = this.timestamp
+            .split(/[^0-9]/, 6)
+            .map(v => parseInt(v, 10))
+            .reverse();
```

Comment:
As commented above, because some of the test files were missing the date and day parts and this would make it fail - that is also why the array is reversed because the end (time stamp) is always there.